### ISSUE TITLE
bottom textInsets does not change the kCTParagraphStyleSpecifierParagraphSpacing attribute

### ISSUE THREAD

**Issue by purrrminator** [TTTAttributedLabel/TTTAttributedLabel#494]
I'm trying to set 30-pt space between paragraphs for string `@"paragraph1\nparagraph2"` with this code:

```
TTTAttributedLabel *footerLabel = [TTTAttributedLabel new];
footerLabel.textInsets = UIEdgeInsetsMake(0, 0, 30, 0);
footerLabel.verticalAlignment = TTTAttributedLabelVerticalAlignmentTop;
footerLabel.lineBreakMode = NSLineBreakByWordWrapping;
footerLabel.numberOfLines = 0;
footerLabel.text = myString;
```
but nothing happens. However, `UIEdgeInsetsMake(30, 0, 0, 0);` does work for me, but this is not what I want, obviously.

Documentation for `textInsets` says:
```
## `UIEdgeInset` Member Correspondence With `CTParagraphStyleSpecifier` Values:
 
 - `top`: `kCTParagraphStyleSpecifierParagraphSpacingBefore`
 - `left`: `kCTParagraphStyleSpecifierHeadIndent`
 - `bottom`: `kCTParagraphStyleSpecifierParagraphSpacing`
 - `right`: `kCTParagraphStyleSpecifierTailIndent`
```

**Comment by jhersh** [comment]
This is a peculiar case. TTTAttributedLabel's `textInsets` documentation states:

> The `UIEdgeInset` members correspond to paragraph style properties rather than a particular geometry, and can change depending on the writing direction. 

However, we can see that:

- `sizeThatFits:` increases the returned `CGSize` by the label's `textInsets`
- `drawTextInRect:` insets all drawn text by the values specified in `textInsets`

These both sound awfully like a "particular geometry", and indeed, the `textInsets` property has no connection to any paragraph style properties. In fact, the label only uses one of the four `CTParagraphStyleSpecifier` values you note above: there is a dedicated `firstLineIndent` property, which corresponds to `kCTParagraphStyleSpecifierHeadIndent` and `NSParagraphStyle`'s `firstLineHeadIndent`.

In short: it appears that the documentation is exactly backwards. I would propose:

- Updating the `textInsets` documentation to note that it really does mean a "particular geometry" and that it insets all drawn text by the values specified. 
- Remove `textInsets` documentation references to paragraph styles
- Update `firstLineIndent` documentation to note it corresponds to `kCTParagraphStyleSpecifierHeadIndent` and the `firstLineHeadIndent` paragraph style
- In another issue, track an enhancement to the label that would add support for the other 3 paragraph specifiers mentioned above, perhaps by adding 3 new label properties to go with the `firstLineIndent` property.

**Comment by cerupcat** [comment]
To support indents in sub-paragraphs, is there a certain place we might look to implement this in the code?

From what I can tell, the firstLineIndent applies to the whole text rather than sub paragraphs within an attributed string. I'm currently using DTCoreText to format html, but using TTAttributedLabel for the label. DTCoreText create sub paragraphs with different indentation depending on the html (like an ordered list for example), but TTAttributedLabel doesn't seem to fully respect the indents.

Here's an example of what DTCoreText gives back for an attributedString

```
We are looking for an SLPA to assist our senior speech therapist by providing direct therapy during afternoon and evening hours. Weekend hours are also available.
 
We will only look at candidates who have:
{
    CTForegroundColor = "<CGColor 0x7f897b0a7240> [<CGColorSpace 0x7f897b215500> (kCGColorSpaceDeviceRGB)] ( 0 0 0 1 )";
    NSFont = "<UICTFont: 0x7f897ad34920> font-family: \"SourceSansPro-Regular\"; font-weight: normal; font-style: normal; font-size: 14.00pt";
    NSParagraphStyle = "<CTParagraphStyle: 0x7f897b0d3c40>{base writing direction = -1, alignment = 4, line break mode = 0, default tab interval = 36\nfirst line head indent = 0, head indent = 0, tail indent = 0\nline height multiple = 0, maximum line height = 0, minimum line height = 0\nline spacing adjustment = 0, paragraph spacing = 14, paragraph spacing before = 0\n}";
}	•	{
    CTForegroundColor = "<CGColor 0x7f897b6cf030> [<CGColorSpace 0x7f897b215500> (kCGColorSpaceDeviceRGB)] ( 0 0 0 1 )";
    DTField = "{listprefix}";
    DTTextLists =     (
        "<DTCSSListStyle 0x7f897b5d7e70 type=5 position=2>"
    );
    NSFont = "<UICTFont: 0x7f897b05bc50> font-family: \"SourceSansPro-Regular\"; font-weight: normal; font-style: normal; font-size: 14.00pt";
    NSParagraphStyle = "<CTParagraphStyle: 0x7f897b05c650>{base writing direction = -1, alignment = 4, line break mode = 0, default tab interval = 100\nfirst line head indent = 0, head indent = 15, tail indent = 0\nline height multiple = 0, maximum line height = 0, minimum line height = 0\nline spacing adjustment = 0, paragraph spacing = 0, paragraph spacing before = 0\ntabs = <CFArray 0x7f897b0d3a40 [0x109698180]>{type = immutable, count = 2, values = (\n\t0 : <CTTextTab: 0x7f897b0cff70>{location = 10, alignment = 1, options = (null)}\n\t1 : <CTTextTab: 0x7f897b0d3880>{location = 15, alignment = 0, options = (null)}\n)}}";
}A current SLPA license
{
    CTForegroundColor = "<CGColor 0x7f897b6cf030> [<CGColorSpace 0x7f897b215500> (kCGColorSpaceDeviceRGB)] ( 0 0 0 1 )";
    DTTextLists =     (
        "<DTCSSListStyle 0x7f897b5d7e70 type=5 position=2>"
    );
    NSFont = "<UICTFont: 0x7f897ad34920> font-family: \"SourceSansPro-Regular\"; font-weight: normal; font-style: normal; font-size: 14.00pt";
    NSParagraphStyle = "<CTParagraphStyle: 0x7f897b6d0ea0>{base writing direction = -1, alignment = 4, line break mode = 0, default tab interval = 36\nfirst line head indent = 15, head indent = 15, tail indent = 0\nline height multiple = 0, maximum line height = 0, minimum line height = 0\nline spacing adjustment = 0, paragraph spacing = 0, paragraph spacing before = 0\n}";
}	{
    CTForegroundColor = "<CGColor 0x7f897b6cf030> [<CGColorSpace 0x7f897b215500> (kCGColorSpaceDeviceRGB)] ( 0 0 0 1 )";
    DTField = "{listprefix}";
    DTTextLists =     (
        "<DTCSSListStyle 0x7f897b5d7e70 type=5 position=2>"
    );
    NSFont = "<UICTFont: 0x7f897b05bc50> font-family: \"SourceSansPro-Regular\"; font-weight: normal; font-style: normal; font-size: 14.00pt";
    NSParagraphStyle = "<CTParagraphStyle: 0x7f897b275b30>{base writing direction = -1, alignment = 4, line break mode = 0, default tab interval = 100\nfirst line head indent = 0, head indent = 15, tail indent = 0\nline height multiple = 0, maximum line height = 0, minimum line height = 0\nline spacing adjustment = 0, paragraph spacing = 14, paragraph spacing before = 0\ntabs = <CFArray 0x7f897b0d36f0 [0x109698180]>{type = immutable, count = 2, values = (\n\t0 : <CTTextTab: 0x7f897b0cff70>{location = 10, alignment = 1, options = (null)}\n\t1 : <CTTextTab: 0x7f897b0d3880>{location = 15, alignment = 0, options = (null)}\n)}}";
}Bilingual skills valued, Spanish preferred
{
    CTForegroundColor = "<CGColor 0x7f897b6cf030> [<CGColorSpace 0x7f897b215500> (kCGColorSpaceDeviceRGB)] ( 0 0 0 1 )";
    DTTextLists =     (
        "<DTCSSListStyle 0x7f897b5d7e70 type=5 position=2>"
    );
    NSFont = "<UICTFont: 0x7f897ad34920> font-family: \"SourceSansPro-Regular\"; font-weight: normal; font-style: normal; font-size: 14.00pt";
    NSParagraphStyle = "<CTParagraphStyle: 0x7f897b275b30>{base writing direction = -1, alignment = 4, line break mode = 0, default tab interval = 100\nfirst line head indent = 0, head indent = 15, tail indent = 0\nline height multiple = 0, maximum line height = 0, minimum line height = 0\nline spacing adjustment = 0, paragraph spacing = 14, paragraph spacing before = 0\ntabs = <CFArray 0x7f897b0d36f0 [0x109698180]>{type = immutable, count = 2, values = (\n\t0 : <CTTextTab: 0x7f897b0cff70>{location = 10, alignment = 1, options = (null)}\n\t1 : <CTTextTab: 0x7f897b0d3880>{location = 15, alignment = 0, options = (null)}\n)}}";
}
```

**Comment by jhersh** [comment]
Ah, it looks like the label's `firstLineIndent` property may not actually correspond to a specific paragraph style. There's definitely more work to do to support all 4 metrics.

**Comment by jhersh** [comment]
As per the comments above, let's move discussion on adding new paragraph metrics support to #561.
